scripts/bfasst/compare_waveforms/file_generation/testbench_generator.py

```diff
@@ -1,110 +1,187 @@
+        if j == total_num(data) - 1:
+            line = f"{line}{total});\n"
+        else:
+            line = f"{line}{total}, "
+    return line
+
+
+def write_inputs(data, tb):
+
+    """Handles writing all of the input signals."""
+
+    for signal_in, bits in zip(data["input_list"], data["input_bits_list"]):
+        signal_in = str(signal_in)
+        bits = str(bits)
+        if signal_in != "clk":
+            line = f"reg [{bits}:0] {signal_in} = 0;\n"
+            tb.write(line)
+        else:
+            line = "reg clk = 0;\n"
+            tb.write(line)
+        line = ""
+    return ""
+
+
+def write_outputs(data, tb):
+
+    """Handles writing all of the output signals."""
+
+    for signal_out, bits in zip(data["output_list"], data["output_bits_list"]):
+        signal_out = str(signal_out)
+        bits = str(bits)
+        line = f"wire [{bits}:0] {signal_out};\n"
+        tb.write(line)
+    return ""
+
+
+def check_clk(data, line):
+    """Confirms that a clock is not declared as input signal and is set to 0."""
+    for input_signal in data["input_list"]:
+        if input_signal == "clk":
+            line = ""
+    return line
+
+
+def write_random_lines(data, test_num, random_list, tb):
+
+    """Writes lines for all randomly generated numbers."""
+
+    # The actual logic for adding random numbers to the testbench.
+    for j in range(int(test_num)):
+        for input_signal, input_number in zip(
+            data["input_list"], range(input_num(data))
+        ):
+            line = write_random_state(input_signal, input_number, j, random_list)
+            tb.write(line)
+            line = ""
+        tb.write("\n")
+    return "    # 5 $finish;"
+
+
+def parse_line(line, data, tb, test_num, paths, random_list):
+
+    """Parses the individual line."""
+
+    if "TB_NAME;" in line:
+        line = write_tb_name(paths, tb, False)
+
+    if "TB_NAME)" in line:
+        line = write_tb_name(paths, tb, True)
+
+    if "INPUTS" in line:
+        line = write_inputs(data, tb)
+
+    if "OUTPUTS" in line:
+        line = write_outputs(data, tb)
+
+    if "MODULE_NAME" in line:
+        line = write_module_name(paths, data)
+
+    if "reg clk = 0;" in line:
+        line = check_clk(data, line)
+
+    if "/*SIGNALS" in line:
+        line = write_random_lines(data, test_num, random_list, tb)
+
+    tb.write(line)
+    return test_num
+
+
 def generate_first_testbench(paths, test_num, data, i):
-    """This function creates the initial testbench that will be modified by the reversed-netlist. It reads in a sample testbench
-    and replaces certain variables with the corresponding information from the data structure. It also sets the variables equal
+
+    """This function creates the initial testbench that will be modified by the reversed-netlist.
+    It reads in a sample testbench
+    and replaces certain variables with the corresponding information from the data structure. It
+    also sets the variables equal
```

Comment:
You could join these lines to make it flow better.  On linux, you can run this block through `fmt` and it will spit out a nicely formatted version.